File: backend/src/main/java/com/belledonne/ecommerce/service/SearchService.java
package com.belledonne.ecommerce.service;

import com.belledonne.ecommerce.entity.SearchAnalytics;
import com.belledonne.ecommerce.entity.SearchSynonym;
import com.belledonne.ecommerce.exception.BadRequestException;
import com.belledonne.ecommerce.exception.ResourceNotFoundException;
import com.belledonne.ecommerce.repository.SearchAnalyticsRepository;
import com.belledonne.ecommerce.repository.SearchSynonymRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;

@Service
@RequiredArgsConstructor
@Transactional
public class SearchService {

    private final SearchAnalyticsRepository searchAnalyticsRepository;
    private final SearchSynonymRepository searchSynonymRepository;

    public void recordSearch(String query) {
        if (query == null || query.isBlank()) {
            return;
        }
        String trimmed = query.trim();
        Optional<SearchAnalytics> existing = searchAnalyticsRepository.findBySearchTermIgnoreCase(trimmed);
        if (existing.isPresent()) {
            SearchAnalytics entry = existing.get();
            entry.setSearchCount(entry.getSearchCount() + 1);
            entry.setLastSearchedAt(LocalDateTime.now());
            searchAnalyticsRepository.save(entry);
        } else {
            SearchAnalytics entry = SearchAnalytics.builder()
                .searchTerm(trimmed)
                .searchCount(1)
                .lastSearchedAt(LocalDateTime.now())
                .build();
            searchAnalyticsRepository.save(entry);
        }
    }

    @Transactional(readOnly = true)
    public String getExpandedTerms(String query) {
        if (query == null || query.isBlank()) {
            return "";
        }
        String trimmed = query.trim().toLowerCase();
        Set<String> termsSet = new LinkedHashSet<>();
        termsSet.add(trimmed); // Always search original term first

        // 1. Exact match checks
        List<SearchSynonym> exactSyns = searchSynonymRepository.findByKeywordIgnoreCase(trimmed);
        for (SearchSynonym syn : exactSyns) {
            termsSet.add(syn.getMappedTerm().toLowerCase());
        }

        // 2. Token-by-token checks for phrases (e.g. "cool fridge" -> "cool refrigerator")
        String[] tokens = trimmed.split("\\s+");
        if (tokens.length > 1) {
            for (String token : tokens) {
                List<SearchSynonym> tokenSyns = searchSynonymRepository.findByKeywordIgnoreCase(token);
                for (SearchSynonym syn : tokenSyns) {
                    String expanded = trimmed.replace(token, syn.getMappedTerm().toLowerCase());
                    termsSet.add(expanded);
                }
            }
        }

        return String.join("|", termsSet);
    }

    // --- Search Analytics Endpoints ---
    @Transactional(readOnly = true)
    public List<SearchAnalytics> getTrendingSearches(int limit) {
        return searchAnalyticsRepository.findByOrderBySearchCountDesc(PageRequest.of(0, limit)).getContent();
    }

    // --- Synonyms CRUD ---
    @Transactional(readOnly = true)
    public List<SearchSynonym> getAllSynonyms() {
        return searchSynonymRepository.findAllByOrderByCreatedAtDesc();
    }

    public SearchSynonym createSynonym(String keyword, String mappedTerm) {
        if (keyword == null || keyword.isBlank() || mappedTerm == null || mappedTerm.isBlank()) {
            throw new BadRequestException("Keyword and mapped term are required");
        }
        String kw = keyword.trim().toLowerCase();
        String mt = mappedTerm.trim().toLowerCase();

        if (searchSynonymRepository.existsByKeywordIgnoreCaseAndMappedTermIgnoreCase(kw, mt)) {
            throw new BadRequestException("Synonym mapping '" + kw + "' -> '" + mt + "' already exists");
        }

        SearchSynonym synonym = SearchSynonym.builder()
            .keyword(kw)
            .mappedTerm(mt)
            .build();
        return searchSynonymRepository.save(synonym);
    }

    public SearchSynonym updateSynonym(Long id, String keyword, String mappedTerm) {
        if (keyword == null || keyword.isBlank() || mappedTerm == null || mappedTerm.isBlank()) {
            throw new BadRequestException("Keyword and mapped term are required");
        }
        SearchSynonym synonym = searchSynonymRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("SearchSynonym", "id", id));

        String kw = keyword.trim().toLowerCase();
        String mt = mappedTerm.trim().toLowerCase();

        // Check duplicate if values changed
        if (!synonym.getKeyword().equalsIgnoreCase(kw) || !synonym.getMappedTerm().equalsIgnoreCase(mt)) {
            if (searchSynonymRepository.existsByKeywordIgnoreCaseAndMappedTermIgnoreCase(kw, mt)) {
                throw new BadRequestException("Synonym mapping '" + kw + "' -> '" + mt + "' already exists");
            }
        }

        synonym.setKeyword(kw);
        synonym.setMappedTerm(mt);
        return searchSynonymRepository.save(synonym);
    }

    public void deleteSynonym(Long id) {
        SearchSynonym synonym = searchSynonymRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("SearchSynonym", "id", id));
        searchSynonymRepository.delete(synonym);
    }
}

File: src/services/searchService.ts
import axiosInstance from '../api/axiosInstance';
import { ENDPOINTS } from '../api/endpoints';

export interface SearchAnalytics {
  id: number;
  searchTerm: string;
  searchCount: number;
  lastSearchedAt: string;
}

export interface SearchSynonym {
  id: number;
  keyword: string;
  mappedTerm: string;
  createdAt: string;
}

let synonymsPromise: Promise<SearchSynonym[]> | null = null;

/**
 * Fetch public list of synonyms and cache the Promise to avoid duplicate hits on search focus.
 */
export const getSynonyms = (force = false): Promise<SearchSynonym[]> => {
  if (!synonymsPromise || force) {
    synonymsPromise = axiosInstance
      .get(ENDPOINTS.SYNONYMS)
      .then((res) => res.data.data || [])
      .catch((err) => {
        console.error('Failed to fetch search synonyms', err);
        synonymsPromise = null;
        return [];
      });
  }
  return synonymsPromise;
};

/**
 * Fetch top trending search terms.
 */
export const getTrendingSearches = async (limit = 10): Promise<SearchAnalytics[]> => {
  const response = await axiosInstance.get(`${ENDPOINTS.SEARCH_TRENDING}?limit=${limit}`);
  return response.data.data || [];
};

/**
 * Admin: Get all synonyms.
 */
export const adminGetSynonyms = async (): Promise<SearchSynonym[]> => {
  const response = await axiosInstance.get(ENDPOINTS.ADMIN_SYNONYMS);
  return response.data.data || [];
};

/**
 * Admin: Add a new synonym mapping.
 */
export const createSynonym = async (keyword: string, mappedTerm: string): Promise<SearchSynonym> => {
  const response = await axiosInstance.post(ENDPOINTS.ADMIN_SYNONYMS, { keyword, mappedTerm });
  // Invalidate public cache
  synonymsPromise = null;
  return response.data.data;
};

/**
 * Admin: Edit an existing synonym mapping.
 */
export const updateSynonym = async (id: number, keyword: string, mappedTerm: string): Promise<SearchSynonym> => {
  const response = await axiosInstance.put(ENDPOINTS.ADMIN_SYNONYM_BY_ID(id), { keyword, mappedTerm });
  // Invalidate public cache
  synonymsPromise = null;
  return response.data.data;
};

/**
 * Admin: Delete a synonym.
 */
export const deleteSynonym = async (id: number): Promise<void> => {
  await axiosInstance.delete(ENDPOINTS.ADMIN_SYNONYM_BY_ID(id));
  // Invalidate public cache
  synonymsPromise = null;
};
